### PR TITLE
Added more settings and results view

### PR DESCRIPTION
**Pull request by alexeichhorn**
I was looking for something like this for at least half a year. Great work.

I added some optional settings to configure the search bar further and also implemented a SwiftUI solution for search results controller to show search results in a separate view. I've tested everything and it seems to work fine for me.
The search results implementation is a bit messy, but I didn't come up with a cleaner solution for now.

I hope these changes help to use this search bar in production since Apple didn't announce a better solution this week at WWDC.

### CODE REVIEW

File: Sources/SwiftlySearch/SwiftlySearch.swift [hunk 1]
```diff
@@ -86,11 +109,25 @@ fileprivate struct SearchBar: UIViewControllerRepresentable {
             }
         }
         
+        var hidesSearchBarWhenScrolling: Bool = true {
+            didSet {
+                self.parent?.navigationItem.hidesSearchBarWhenScrolling = hidesSearchBarWhenScrolling
+            }
+        }
```
Comment (thislooksfun) [inline Sources/SwiftlySearch/SwiftlySearch.swift:116]: I don't see why this is necessary. Could you explain why this is needed in addition to line 127?

File: Sources/SwiftlySearch/SwiftlySearch.swift [hunk 2]
```diff
@@ -53,16 +69,23 @@ fileprivate struct SearchBar: UIViewControllerRepresentable {
...
             super.init()
             
             searchController.searchResultsUpdater = self
-            searchController.hidesNavigationBarDuringPresentation = true
+            searchController.hidesNavigationBarDuringPresentation = hidesNavigationBarDuringPresentation
             searchController.obscuresBackgroundDuringPresentation = false
             
+            if let placeholder = placeholder {
+                searchController.searchBar.placeholder = placeholder
+            }
```
Comment (thislooksfun) [inline Sources/SwiftlySearch/SwiftlySearch.swift:87]: Since both `placeholder` and `searchController.searchBar.placeholder` are of type `String?`, I think this `if let` is redundant.

```suggestion
            searchController.searchBar.placeholder = placeholder
```

File: Sources/SwiftlySearch/SwiftlySearch.swift [hunk 3]
```diff
@@ -53,16 +69,23 @@ fileprivate struct SearchBar: UIViewControllerRepresentable {
...
         private var subscription: AnyCancellable?
         
-        init(text: Binding<String>) {
+        init(text: Binding<String>, placeholder: String?, hidesNavigationBarDuringPresentation: Bool, resultContent: (String) -> ResultContent?) {
             self._text = text
-            self.searchController = UISearchController(searchResultsController: nil)
+            
+            let resultView = resultContent(text.wrappedValue)
+            let searchResultController = resultView.map { UIHostingController(rootView: $0) }
+            self.searchController = UISearchController(searchResultsController: searchResultController)
```
Comment (thislooksfun) [inline Sources/SwiftlySearch/SwiftlySearch.swift:77]: I have not had a chance to test this out yet, but could you please explain the rationale behind this code? What is the expected use case?

File: Sources/SwiftlySearch/SwiftlySearch.swift [hunk 4]
```diff
@@ -21,17 +21,29 @@ import SwiftUI
...
 public extension View {
-    func navigationBarSearch(_ searchText: Binding<String>) -> some View {
-        return overlay(SearchBar(text: searchText).frame(width: 0, height: 0))
+    func navigationBarSearch(_ searchText: Binding<String>, placeholder: String? = nil, hidesNavigationBarDuringPresentation: Bool = true, hidesSearchBarWhenScrolling: Bool = true) -> some View {
+        return overlay(SearchBar<AnyView>(text: searchText, placeholder: placeholder, hidesNavigationBarDuringPresentation: hidesNavigationBarDuringPresentation, hidesSearchBarWhenScrolling: hidesSearchBarWhenScrolling).frame(width: 0, height: 0))
+    }
+    
+    func navigationBarSearch<ResultContent: View>(_ searchText: Binding<String>, placeholder: String? = nil, hidesNavigationBarDuringPresentation: Bool = true, hidesSearchBarWhenScrolling: Bool = true, @ViewBuilder resultContent: @escaping (String) -> ResultContent) -> some View {
+        return overlay(SearchBar(text: searchText, placeholder: placeholder, hidesNavigationBarDuringPresentation: hidesNavigationBarDuringPresentation, hidesSearchBarWhenScrolling: hidesSearchBarWhenScrolling, resultContent: resultContent).frame(width: 0, height: 0))
     }
```
Comment (thislooksfun) [inline Sources/SwiftlySearch/SwiftlySearch.swift:30]: Is there a way to make these into a single function definition instead of having to rely on overloading?

Comment (alexeichhorn) [reply Sources/SwiftlySearch/SwiftlySearch.swift:30]: I tried some things, but couldn't find a good solution. It seems like overloading is the best bet for handling optional generics. (e.g. discussed here: [Stackoverflow](https://stackoverflow.com/questions/28458160/default-value-for-optional-generic-parameter-in-swift-function))

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 2 above)
Comment (alexeichhorn) [reply Sources/SwiftlySearch/SwiftlySearch.swift:87]: I thought that too, but search bars actually don't have a default value of nil. Their default value is the localized search title. Setting it to nil will display an empty string.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 1 above)
Comment (alexeichhorn) [reply Sources/SwiftlySearch/SwiftlySearch.swift:116]: I'm not 100% sure how SwiftUI manages `UIViewControllerRepresentable` objects and their view controller. But as I understand it, one `SearchBarWrapperController` object can be handled by multiple `SearchBar` objects during its lifecycle. Therefore `hidesSearchBarWhenScrolling` can be changed and has to be somehow updated. 
Probably not something a lot of people would do, but it's kind of SwiftUI's paradigm to be able to update anything.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 3 above)
Comment (alexeichhorn) [reply Sources/SwiftlySearch/SwiftlySearch.swift:77]: The use case is the same as in UIKit. You can use this for displaying search results in a different context than the view below. It can also be used for example for search suggestions. 
I'm not sure, if I understood your question correctly though.

**Comment by alexeichhorn** [comment]
@thislooksfun Thanks for your valid concerns. I hope my answers are useful to you.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 2 above)
Comment (thislooksfun) [reply Sources/SwiftlySearch/SwiftlySearch.swift:87]: Ah, yeah that makes sense. Fair enough.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 4 above)
Comment (thislooksfun) [reply Sources/SwiftlySearch/SwiftlySearch.swift:30]: That's a shame. Ah well. Maybe in a future version of Swift there will be a better solution.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 1 above)
Comment (thislooksfun) [reply Sources/SwiftlySearch/SwiftlySearch.swift:116]: Yeah that makes sense. I also forgot I had a similar fix on `searchController` since the diff hid those lines.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 3 above)
Comment (thislooksfun) [reply Sources/SwiftlySearch/SwiftlySearch.swift:77]: I was worried how it would behave if there *wasn't* a result view, but having tested it it indeed works as I wanted it to.

File: Sources/SwiftlySearch/SwiftlySearch.swift [hunk 5]
```diff
@@ -86,11 +109,25 @@ fileprivate struct SearchBar: UIViewControllerRepresentable {
...
         override func viewDidAppear(_ animated: Bool) {
+            setup()
+        }
+        
+        private func setup() {
             self.parent?.navigationItem.searchController = searchController
+            self.parent?.navigationItem.hidesSearchBarWhenScrolling = hidesSearchBarWhenScrolling
+            
+            // make search bar appear at start (default behaviour since iOS 13)
+            self.parent?.navigationController?.navigationBar.sizeToFit()
```
Comment (thislooksfun) [inline Sources/SwiftlySearch/SwiftlySearch.swift:130]: This does not seem to do anything that I could tell. Why was this added?

**Comment by thislooksfun** [comment]
@alexeichhorn Sorry for the delay, the last couple of weeks were much busier than expected. I have just one more question (above) before I sign off on this.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 5 above)
Comment (alexeichhorn) [reply Sources/SwiftlySearch/SwiftlySearch.swift:130]: Without this line of code, the search bar wouldn't show up when the view first appears and `hidesSearchBarWhenScrolling` is true. The default behavior (since iOS 13) in UIKit is to show the search bar initially, only hiding it when the user interacts with the scroll view. But sometimes there's a bug in UIKit, which lets the search bar disappear initially. So this is basically a bug fix until Apple fixes this bug.

**Comment by alexeichhorn** [comment]
@thislooksfun I posted my answer.

File: Sources/SwiftlySearch/SwiftlySearch.swift (hunk 5 above)
Comment (thislooksfun) [reply Sources/SwiftlySearch/SwiftlySearch.swift:130]: Alright, that's what I figured it was supposed to do, but I was unable to reproduce that for some reason (namely it hid itself by default even with that line).

**Comment by thislooksfun** [comment]
Published in [v1.1.0](https://github.com/thislooksfun/SwiftlySearch/releases/tag/v1.1.0)!
